powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/SignatureUtils.java

```diff
@@ -98,21 +98,27 @@ public boolean validateECDSASignature(byte[] signedBytes, byte[] signature, Publ
      * @param data Data to be signed.
      * @param signatureKeys Keys for computing the signature.
      * @param ctrData Counter byte array / derived key index.
+     * @param length Required length of the factor related signature component (i.e, if length is 4, then 2FA will
+     *               have 8 digits). Minimal allowed non-null value is 4. If the provided value is less than 4, the value
+     *               is increased automatically to 4. Maximum allowed value is 12, values above are again set to 12.
+     *               If the value is null, the default system value (8) is used.
      * @return Decimal formatted PowerAuth signature for given data.
      * @throws GenericCryptoException In case signature computation fails.
      * @throws CryptoProviderException In case cryptography provider is incorrectly initialized.
      */
-    private String computePowerAuthDecimalSignature(byte[] data, List<SecretKey> signatureKeys, byte[] ctrData) throws GenericCryptoException, CryptoProviderException {
+    private String computePowerAuthDecimalSignature(byte[] data, List<SecretKey> signatureKeys, byte[] ctrData, Integer length) throws GenericCryptoException, CryptoProviderException {
         // Prepare holder for signature components
         final String[] signatureStringComponents = new String[signatureKeys.size()];
         // Compute signature components
         final List<byte[]> signatureComponents = computePowerAuthSignatureComponents(data, signatureKeys, ctrData);
+        // Determine the length of the signature component
+        int signatureDecimalLength = (length == null) ? PowerAuthConfiguration.SIGNATURE_DECIMAL_LENGTH : Math.min(Math.max(length, 4), 12);
         // Convert byte components into decimal signature
         for (int i = 0; i < signatureComponents.size(); i++) {
             final byte[] signatureComponent = signatureComponents.get(i);
             int index = signatureComponent.length - 4;
-            int number = (ByteBuffer.wrap(signatureComponent).getInt(index) & 0x7FFFFFFF) % (int) (Math.pow(10, PowerAuthConfiguration.SIGNATURE_DECIMAL_LENGTH));
-            signatureStringComponents[i] = String.format("%0" + PowerAuthConfiguration.SIGNATURE_DECIMAL_LENGTH + "d", number);
+            int number = (ByteBuffer.wrap(signatureComponent).getInt(index) & 0x7FFFFFFF) % (int) (Math.pow(10, signatureDecimalLength));
```

Comment:
Several lines above you set the upper limit to 12 which will cause integer to overflow here. 10^12 is way bigger than 2^31.

Reply:
Good point.